sys/shell/shell.c

```diff
@@ -45,171 +46,191 @@ static void _putchar(int c) {
+{
+    shell_command_handler_t handler = NULL;
 
-        const shell_command_t *entry;
+    if (command_list != NULL) {
+        handler = search_commands(command_list, command);
+    }
 
-        if ((entry = command_lists[i])) {
-            /* iterating over commands in command_lists entry */
-            while (entry->name != NULL) {
-                if (strcmp(entry->name, command) == 0) {
-                    return entry->handler;
-                }
-                else {
-                    entry++;
-                }
-            }
-        }
+    if (handler == NULL && _builtin_cmds != NULL) {
+        handler = search_commands(_builtin_cmds, command);
     }
 
-    return NULL;
+    return handler;
 }
 
-static void print_help(const shell_command_t *command_list)
+static void print_commands(const shell_command_t *entry)
 {
-    printf("%-20s %s\n", "Command", "Description");
-    puts("---------------------------------------");
-
-    const shell_command_t *command_lists[] = {
-        command_list,
-#ifdef MODULE_SHELL_COMMANDS
-        _shell_command_list,
-#endif
-    };
+    for (; entry->name != NULL; entry++) {
+        printf("%-20s %s\n", entry->name, entry->desc);
+    }
+}
 
-    /* iterating over command_lists */
-    for (unsigned int i = 0; i < ARRAY_SIZE(command_lists); i++) {
+static void print_help(const shell_command_t *command_list)
+{
+    puts("Command              Description\n---------------------------------------");
 
-        const shell_command_t *entry;
+    if (command_list != NULL) {
+        print_commands(command_list);
+    }
 
-        if ((entry = command_lists[i])) {
-            /* iterating over commands in command_lists entry */
-            while (entry->name != NULL) {
-                printf("%-20s %s\n", entry->name, entry->desc);
-                entry++;
-            }
-        }
+    if (_builtin_cmds != NULL) {
+        print_commands(_builtin_cmds);
     }
 }
 
+enum PARSE_STATE {
+    PARSE_SPACE,
+    PARSE_UNQUOTED,
+    PARSE_SINGLEQUOTE,
+    PARSE_DOUBLEQUOTE,
+    PARSE_ESCAPE_MASK,
+    PARSE_UNQUOTED_ESC,
+    PARSE_SINGLEQUOTE_ESC,
+    PARSE_DOUBLEQUOTE_ESC,
+};
+
+static enum PARSE_STATE escape_toggle(enum PARSE_STATE s)
+{
+    return s ^ PARSE_ESCAPE_MASK;
+}
+
+#define SQUOTE '\''
+#define DQUOTE '"'
+#define ESCAPECHAR '\\'
+#define BLANK ' '
+
+/**
+ * Break input line into words, create argv and call the command handler.
+ *
+ * Words are broken up at spaces. A backslash escaped the character that comes
+ * after (meaning it it taken literally and if it is a space it does not break
+ * the word). Spaces can also be protected by quoting with double or single
+ * quotes.
+ *
+ State diagram for the tokenizer:
+```
+
+
+           ┌───[\]────┐   ┌─────["]────┐   ┌───[']─────┐  ┌───[\]────┐
+           ↓          │   ↓            │   │           ↓  │          ↓
+  ┏━━━━━━━━━━┓      ┏━┷━━━━━┓        ┏━┷━━━┷━┓       ┏━━━━┷━━┓     ┏━━━━━━━━━━┓
```

Comment:
This displays badly because github's font is defective. In your editor or terminal it should display fine.